app/services/search_api_service.rb

```diff
@@ -0,0 +1,45 @@
+class SearchApiService
+
+  def initialize(attrs = {})
+    @type = attrs[:type].downcase
+    @search_word = attrs[:q]
+  end
+
+  def search
+    if search_word
+      search_word_search
+    else
+      all_search
+    end
+  end
+
+  private
+    attr_reader :type, :search_word
+
+    def all_search
+      case type
+      when 'items', 'item'
+        Item.all
+      when 'stores', 'store'
+        Store.all
+      when 'users', 'user'
+        User.all
+      when 'category', 'categories'
+        Category.all
+      end
+    end
+
+    def search_word_search
+      insensitive_search_word = search_word.downcase
+      case type
+      when 'items', 'item'
```

Comment:
This is interesting, maybe after the project you could explain more of what's happening in the code. Simply so I could learn from it.

Reply:
Absolutely!

I was trying to find a quick way to take in the query from the params and make it lowercase to before doing the "ILIKE", or insensitive likeness, query on the columns.

I wanted to make the query word optional, so when I tried adding the downcase method call in the initialize, it was throwing a nil error if the search word didn't exist

Documentation here has more information https://www.postgresql.org/docs/7.3/static/functions-matching.html
: "The keyword ILIKE can be used instead of LIKE to make the match case insensitive according to the active locale. This is not in the SQL standard but is a PostgreSQL extension."